i2c/mcp9808_i2c/mcp9808_i2c.c

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include "pico/stdlib.h"
+#include "pico/binary_info.h"
+#include "hardware/i2c.h"
+
+/* Example code to talk to a MCP9808 ±0.5°C Digital temperature Sensor
+   
+   This reads and writes to registers on the board. 
+
+   Connections on Raspberry Pi Pico board, other boards may vary.
+
+   GPIO PICO_DEFAULT_I2C_SDA_PIN (On Pico this is GP4 (physical pin 6)) -> SDA on MCP9808 board
+   GPIO PICO_DEFAULT_I2C_SCK_PIN (On Pico this is GP5 (physcial pin 7)) -> SCL on MCP9808 board
+   Vsys (physical pin 39) -> VDD on MCP9808 board
+   GND (physical pin 38)  -> GND on MCP9808 board
+
+*/
+//The bus address is determined by the state of pins A0, A1 and A2 on the MCP9808 board
+static uint8_t ADDRESS = 0x18;
+
+const uint8_t WRITE_MODE = 0xFE;
+const uint8_t READ_MODE = 0xFF;
+
+
+//hardware registers
+
+const uint8_t REG_POINTER = 0x00;
+const uint8_t REG_CONFIG = 0x01;
+const uint8_t REG_TEMP_UPPER = 0x02;
+const uint8_t REG_TEMP_LOWER = 0x03;
+const uint8_t REG_TEMP_CRIT = 0x04;
+const uint8_t REG_TEMP_AMB = 0x05;
+const uint8_t REG_MAN_ID = 0x06;
+const uint8_t REG_DEV_ID = 0x07;
+const uint8_t REG_RESOLUTION = 0x08;
+
+
+void mcp9808_check_limits(uint8_t upper_byte) {
+
+    // Check flags and raise alerts accordingly 
+    if ((upper_byte & 0x40) == 0x40){ //TA > TUPPER
+        printf("Temperature is above the upper temperature limit.\n");
+    }
+    if ((upper_byte & 0x20) == 0x20) { //TA < TLOWER
+        printf("Temperature is below the lower temperature limit.\n");
+    }
+    if ((upper_byte & 0x80) == 0x80) { //TA > TCRIT
+        printf("Temperature is above the critical temperature limit.\n");
+    }
+}
+
+float mcp9808_convert_temp(uint8_t upper_byte, uint8_t lower_byte) {
+
+    float temperature;
+
+
+    //Check if TA <= 0°C and convert to denary accordingly
+    if ((upper_byte & 0x10) == 0x10) { 
+        upper_byte = upper_byte & 0x0F; 
+        temperature = 256 - (((float)upper_byte * 16) + ((float)lower_byte / 16));
+    } else {
+        temperature = (((float)upper_byte * 16) + ((float)lower_byte / 16));
+  
+    }
+    return temperature;
+}
+
+
+void mcp9808_set_limits() {
+    
+    //Set an upper limit of 30°C for the temperature
+    uint8_t upper_temp_msb = 0x01;
+    uint8_t upper_temp_lsb = 0xE0;
+
+    //Set a lower limit of 20°C for the temperature
+    uint8_t lower_temp_msb = 0x01;
+    uint8_t lower_temp_lsb = 0x40;
+
+    //Set a critical limit of 40°C for the temperature
+    uint8_t crit_temp_msb = 0x02;
+    uint8_t crit_temp_lsb = 0x80;
+
+    uint8_t buf[3];
+    buf[0] = REG_TEMP_UPPER;
+    buf[1] = upper_temp_msb;
+    buf[2] = upper_temp_lsb; 
+    i2c_write_blocking(i2c_default, ADDRESS, buf, 3, false);

```

Comment:
I think the use of `i2c_default` here _might_ mean that the whole of this `mcp9808_set_limits()` function needs to be inside a `#ifdef i2c_default ... #endif` guard? Graham will be able to confirm/deny.

Reply:
Yep, this is what I gathered as well.